lib/spack/llnl/util/symlink.py

```diff
@@ -340,3 +354,87 @@ class SymlinkError(SpackError):
     """Exception class for errors raised while creating symlinks,
     junctions and hard links
     """
+
+
+class SymlinkAlreadyExistsError(SymlinkError):
+    """Link path already exists."""
+
+
+class SymlinkSourceAbsButNotExistError(SymlinkError):
+    """Source path is absolute but does not exist."""
+
+
+class SymlinkSourceNotRelativeToLinkError(SymlinkError):
+    """The source path is not relative to the link path."""
+
+
+class NoWinCreateLinkOnNixError(SymlinkError):
+    """windows_create_link method can't be used on non-Windows OS."""
+
+
+class SymlinkNotFileOrDirError(SymlinkError):
+    """Cannot create link. It is neither a file nor a directory."""
+
+
+class NoJunctionOnNixError(SymlinkError):
+    """windows_create_junction method can't be used on non-Windows OS."""
+
+
+class NoSourcePathForJunctionError(SymlinkError):
+    """Source path does not exist, cannot create a junction."""
+
+
+class JunctionLinkAlreadyExistsError(SymlinkError):
+    """Link path already exists. Cannot create hard link."""
+
+
+class JunctionSourceNotADirError(SymlinkError):
+    """Source path is not a directory, cannot create a junction."""
+
+
+class JunctionCreationError(SymlinkError):
+    """Make junction command returned a non-zero return code."""
+
+
+class NoHardlinkOnNixError(SymlinkError):
+    """windows_create_hard_link method can't be used on non-Windows Os."""
+
+
+class NoFileForHardlinkError(SymlinkError):
+    """File path does not exist. Cannot create hard link."""
+
+
+class HardlinkPathAlreadyExistsError(SymlinkError):
+    """Link path already exists. Cannot create hard link."""
+
+
+class HardlinkPathNotAFileError(SymlinkError):
+    """File path is not a file. Cannot create hard link."""
+
+
+class CantReadHardlinkOnNixError(SymlinkError):
+    """Can't read hard link on non-Windows OS."""
+
+
+class HardlinkReadError(SymlinkError):
+    """An error occurred while reading hard link."""
+
+
+class MultiPathsForinodeError(SymlinkError):
+    """Found multiple paths pointing to the same inode."""
+
+
+class NoHardlinkSourcePathError(SymlinkError):
+    """Cannot determine hard link source path."""
+
+
+class CantReadJunctionOnNixError(SymlinkError):
+    """Can't read junction on non-Windows OS."""
+
+
+class JunctionReadError(SymlinkError):
+    """An error occurred while reading junction."""
+
+
+class NoJunctionPathError(SymlinkError):
+    """Could not find junction path."""
```

Comment:
I think we should create new exception types only if we handle them. Namely the only 3 we handle in this PR are

```
            except llnl.util.symlink.SymlinkAlreadyExistsError:
                report_already_linked()
            except llnl.util.symlink.JunctionLinkAlreadyExistsError:
                report_already_linked()
            except llnl.util.symlink.HardlinkPathAlreadyExistsError:
                report_already_linked()
```

I was originally thinking these 3 could be consolidated under

```
class AlreadyExists(SymlinkError):
    pass
```

such that you need only catch one. Apologies for not explaining myself well here. Do you see a reason to have the additional exception types?

Reply:
I want to avoid creating exception classes that we have no special handling for. IMO an exception type without any handling isn't very useful (and therefore is just taking up space).

Reply:
Sounds good! I'll go with the singular already exists error.

Reply:
Addressed and pushed